src/lcd_st7789v.c

```diff
@@ -0,0 +1,188 @@
+// Commands for sending messages to an st7789V display connected
+// using the "8080-II series" interface
+//
+// Copyright (C) 2020  Martijn van Buul <martijn.van.buul@gmail.com>
+// Copyright (C) 2018  Kevin O'Connor <kevin@koconnor.net>
+//
+// This file may be distributed under the terms of the GNU GPLv3 license.
+
+#include "autoconf.h" // CONFIG_CLOCK_FREQ
+#include "basecmd.h" // oid_alloc
+#include "board/gpio.h" // gpio_out_write
+#include "board/irq.h" // irq_disable
+#include "board/misc.h" // timer_from_us
+#include "command.h" // DECL_COMMAND
+#include "sched.h" // DECL_SHUTDOWN
+
+struct st7789v {
+    uint8_t last_dcx;
+    uint8_t last_data;
+    struct gpio_out dcx, wrx, databits[8];
+};
+
+
+/****************************************************************
+ * Transmit functions
+ ****************************************************************/
+
+// Write 8 bits to the st7789v using the "8080-II series" interface
+static inline void
+st7789v_xmit_byte(struct st7789v *h, uint8_t data, uint8_t dcx )
+{
+    uint8_t mask = 0x01;
+    gpio_out_write(h->wrx, 0);
+    if (h->last_dcx != dcx)
+    {
+        gpio_out_write(h->dcx, dcx);
+        h->last_dcx = dcx;
+    }
+
+    for (int bit = 0; bit < 8; ++bit, mask <<=1)
+    {
+        const uint8_t new_data = data & mask;
+        if ( new_data != (h->last_data & mask) )
+        {
+            gpio_out_write( h->databits[bit], new_data);
+        }
+    }
+
+    gpio_out_write(h->wrx, 1);
+    h->last_data = data;
+}
+
+// Transmit a series of bytes to the chip. The first byte in every
+// request is assumed to be the 8-bit command; subsequent bytes
+// are parameters or display data.
+static void
+st7789v_xmit(struct st7789v *h, uint8_t len, uint8_t *data)
+{
+    for (uint8_t i = 0; i < len; ++i) {
+        uint8_t b = *data++;
+        st7789v_xmit_byte(h, b, (i != 0) );
+    }
+ }
+
+// Flood-fill bitmap data. This assumes 16bpp pixel format
+static void
+st7789v_floodfill(struct st7789v *h, uint8_t continuation,
+                  uint32_t len, uint16_t data)
+{
+    // RAMWR or WRMEMC/RAMWRC command
+    st7789v_xmit_byte(h, continuation ? 0x3c : 0x2c, 0);
+
+    const uint8_t loByte = data & 0xff;
+    const uint8_t hiByte = data >> 8;
+    for (uint32_t i = 0; i < len; ++i) {
+        st7789v_xmit_byte(h, hiByte, 1 );
+        st7789v_xmit_byte(h, loByte, 1 );
+    }
+}
+
+// Send 1bpp bitmap as 16bpp image data
+static void
+st7789v_bitmap(struct st7789v *h, uint8_t continuation, uint16_t fgcolor,
+               uint16_t bgcolor, uint8_t len, uint8_t *data)
```

Comment:
FWIW, I think you could introduce a simple run-length-encoding (RLE) scheme and replace the three update methods with just one.

Something like:
```C
static void
st7789v_update(struct st7789v *h, uint8_t len, uint8_t *data)
{
    uint8_t *end = data[len], *p = data;
    while (p < data) {
        uint8_t code = *p, literals = code >> 4, copies = code & 0xf;
        if (p + literals > end)
            shutdown("Invalid update");
        while (copies--) {
            uint_fast8_t i;
            for (i=0; i<literals; i++)
                st7789v_xmit_byte(h, p[i], 1);
            }
        }
        p += literals;
    }
}
```

A simple rle scheme (with variable number of literals) should work well for both uncompressable data (as in `st7789v_xmit_byte`), with highly compressable data (as in `st7789v_floodfill`), and with typical fonts (as in `st7789v_bitmap`).

Reply:
I considered that, but there's a nasty catch: This is a hybrid 16/8 bit device. While the low-level bus protocol is byte-sized (it is always a one-byte command followed by byte-wise data, at least in the 8080 protocol employed here) , the framebuffer data itself is *16-bit* wide. This has three annoying consequences:

- There is no bytewise repetition for bitmap data, unless you limit yourself to colors with equal most significant byte and lowest significant byte. If you don't, you immediately hit worst-case behaviour since all your RLE 'runs' will be 1 byte wide only by definition. The obvious fix would be to use RLE using 16-bit integer quantities..
- ... but that would imply that all other commands would have to be written using 16-bit data words as well. If these 16-bit words word be written in big-endian format, this would make sense for some commands(RASET and CASET come to mind, which are used frequently and have 16-bit arguments), but other commands only need 1-byte data. The datasheet is mum about whether it is allowed to supply more arguments than required (and I haven't tried it) (and the single byte of payload would have to end up in the most significant byte of the payload word, but that's a minor inconvenience)  
- Framebuffer data is in RGB565 format - there is no palette which can be used to help reduce the VLQ encoding length; unless you limit yourself to colors without any significant red component,  every pixel value needs *3* bytes to be encoded, plus at least one byte for the run length. This makes it rather inefficient for typical font data (but it would remove the need for the floodfill, which is probably not worth the effort anyway. It's only used during initialisation.). 

Klipper's display is monochromatic (and the framebuffer in klippy is monochromatic). By encoding any bitmap as 1bpp, I need between 40 (320/8 = 40 bytes before encoding, no MSB set)  and 80 (320/8 = 40 bytes before encoding, all with MSB set). If I transfer bitmaps based on 16-bpp RLE encoded data, this would correspond to only 10-20 transitions per line - which is not a lot. In other words: I haven't checked it, but I doubt there's going to be a real-life gain. 

RLE encoding the 1bpp stuffed data makes sense, but that still makes bitmap data stand out from other data. 

I'll gather some statistics, implement the alternatives, do some benchmarking and report back here (might take a while)

Reply:
If you use an RLE scheme with a "literal count" on each RLE command then it should be able to handle 1 or 2 byte updates.  (In my example code above, the format is `<code1><data1><code2><data2>...` where `<code>` is `<4 bit literal count><4 bit copy count>`.  I'm sure other encodings could also work.

Reply:
Just to add a more concrete example, the idea is that the sequence "abcdababab" could be encoded with "<41>abcd<23>ab".

Reply:
> there is no palette which can be used to help reduce the VLQ encoding length

FYI, the VLQ encoding is only done on named parameters of a command.  The content of "%*s" buffers is not encoded - it's contents are copied verbatim.  So, if you have a 20byte binary buffer that is passed as a parameter to a "%*s" field then it will use 20 bytes (plus one byte to pass the length) in the generated message.